### PR TITLE
#fixed Assert selected database atomically for queries

### PR DESCRIPTION
**Pull request by Jason-Morcos**
## Changes:

- Added SPMySQL query overloads that assert the expected database under the same connection lock immediately before query execution, closing the race between `USE` and a dependent query on the shared connection.
- Extended explicit database context through document-scoped custom-query, table, metadata, import/export, copy/rename, structure, index, relation, and trigger paths while retaining non-asserting APIs for genuinely session/server-wide or fully qualified operations.
- Added Swift-owned per-connection assertion state, live client-charset lookup, UTF-8 fallback selection with caller-managed charset restoration, reconnect invalidation, and precise `USE`/`DROP DATABASE` state transitions.
- Updated SQL import and custom-query batches to follow successful standalone `USE` and `DROP DATABASE` statements, including trailing comments, executable comments, server-version/vendor gates, case-sensitive and case-insensitive database names, and quoted identifiers without intervening whitespace.
- Added database-scoped metadata caching and snapshot propagation so metadata and row queries from one operation cannot silently cross schemas.

## Closes following issues:

- Closes: N/A
- Refs discussion #2487

## Tested:

- Processors:
  - [ ] Intel
  - [x] Apple Silicon
- macOS Versions:
  - [ ] 12.x (Monterey)
  - [ ] 13.x (Ventura)
  - [ ] 14.x (Sonoma)
  - [ ] 15.x (Sequoia)
  - [x] Other: 26.5.2
- Localizations:
  - [x] English
  - [ ] Spanish
  - [ ] Other
- Xcode Version:
  Xcode 26.6 (17F113)

Exact-head validation on `ee95652d8`, merged with `main` at `15580e31e`:

- Sequel Ace `Unit Tests`: 757 passed, 60 expected environment-dependent skips, 0 failures (817 total), rerun with isolated DerivedData.
- SPMySQL framework suite: 28 passed, 0 skipped, 0 failures.
- Release configuration build: succeeded with 0 errors. The review-discovered CSV/XML exporter shadow warnings were removed and the build rechecked.
- Current-main compatibility: retained both this PR's database-cache app/test memberships and `main`'s MCP favorite-helper app/test memberships; the combined tree passed full local validation and all required GitHub checks.
- Live MySQL 8.0.46 regressions passed for caller-managed charset restoration, disabled session tracking, concurrent database selection, a multi-statement custom-query batch that follows `USE`, quoted `USE`/`DROP` syntax without whitespace, and copying an unqualified view body into another database so it binds to the target schema.
- `git diff --check`, clean worktree, current-main merge base, GPG signatures, and cleanup of all temporary test schemas were verified.

## Screenshots:

N/A — no UI changes.

## Additional notes:

Root cause: the selected database is sticky server-side, while the old flow selected a database and ran its dependent query in separate lock acquisitions. Another thread could change the connection's database between those operations. This PR makes the assertion and query one atomic locked operation and propagates the intended database to unqualified document/table queries.

The non-asserting methods remain intentionally available. Session setup, server variables, process/user management, database-level DDL, and fully qualified cross-database or information-schema queries do not all have a meaningful selected-database precondition. The enforceable rule is that unqualified document/table operations carry an explicit database context.

### Kaspik review follow-up

- Large custom-query/import statements no longer run the full comment stripper just to prove they are ordinary DML. Both app and framework parsers use an exact, identifier-boundary-aware `USE`/`DROP` prefix guard, with conservative handling only when leading comments require inspection.
- An explicit no-database assertion now reports `A database is unexpectedly selected on this connection.` rather than the misleading server text `No database selected`.
- Invented MariaDB bracket-quoted identifier support was removed. MySQL/MariaDB quote handling is limited to actual supported forms.
- The app and framework comment strippers now cross-reference one another and have mirrored regressions for executable-comment version/vendor gates, overflowed gates, quoted content, and `--` at EOF. Sharing one compiled implementation is impractical here because the app Unit Tests target compiles app sources without the app bridging header, while the framework owns the MySQL C bridge.
- MySQL's accepted no-whitespace quoted forms (``USE`db```, ``DROP DATABASE`db```, and ``DROP DATABASE IF EXISTS`db```) are tracked rather than leaving stale assertion state.
- The cross-database view-copy path is exercised against live MySQL. `SHOW CREATE VIEW` returns an unqualified body; creation is asserted in the target database and the regression proves the copied view reads the target table, not the identically named source table. The ` AS ` delimiter is also used so extraction does not stop on an incidental `AS` substring.
- `SELECT @@lower_case_table_names` is no longer issued at the start of every import/custom-query action. It is loaded at most once, immediately before a case-only `DROP DATABASE`/`DROP SCHEMA` ambiguity; exact and clearly unrelated names need no lookup.
- `BUILD_LIBRARY_FOR_DISTRIBUTION` is retained deliberately. Removing it while using `@_implementationOnly import MySQLClient` produces Xcode's library-evolution instability warning; the setting is required for this framework layout, not copied project churn. The unrelated framework root-group rename was fully removed.
- The `taskArguments` construction change also fixes the latent nil-callback truncation caused by `dictionaryWithObjectsAndKeys:`; this incidental correction is now documented.
- A final build-diagnostic audit found and removed nine new CSV/XML exporter ivar-shadow warnings.
- Successful `USE` and `DROP DATABASE` statements now update the framework's known database state precisely instead of blindly invalidating it. This prevents the next asserted diagnostic query from issuing a hidden context probe that erases warnings; a live `DROP DATABASE IF EXISTS missing_schema; SHOW WARNINGS` regression verifies warning 1008 survives.
- Separate custom-query actions for `ROW_COUNT()`, `FOUND_ROWS()`, and `SHOW WARNINGS` no longer run a metadata query first. A live MySQL regression verifies all three retain diagnostics from the preceding user statement.

The database-level copy behavior and both batch paths are covered automatically and ran against live MySQL. A literal UI click-through and a live MariaDB server were not available in this environment; MariaDB executable-comment/vendor behavior is covered by deterministic mirrored tests.

### CODE REVIEW

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@codex review

File: Frameworks/SPMySQLFramework/SPMySQL Unit Tests/SPMySQLStringAdditions_Tests.m [hunk 1]
```diff
@@ -18,6 +19,360 @@ - (void)test_stringForDataBytesLengthEncoding;
 
 @end
 
+@interface SPMySQLConnectionDatabaseAssertion_Tests : XCTestCase <SPMySQLStreamingResultStoreDelegate>
```
[the remark this answers is not in the document]
Comment (Jason-Morcos) [reply Frameworks/SPMySQLFramework/SPMySQL Unit Tests/SPMySQLStringAdditions_Tests.m]: Addressed in signed commit 3ef4c1144. The live database-assertion regression now lives in SADatabaseAssertionTests.swift, the previous Objective-C test additions are restored to main, and the broader language-policy sweep moved the new database-scoped cache state and coverage to Swift as well.

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@Kaspik Would love a thorough review of this one - this is aiming to once and for all solve the "wrong database" issue reports

**Comment by Kaspik** [comment]
@codex Review every aspect of this PR thoroughly. Challenge every change that might cause an issue and revalidate it. Think deeply about the problem it's trying to solve.

**Comment by Jason-Morcos** [comment]
@codex review

**Review by Kaspik — CHANGES_REQUESTED** [review]
# Review — Assert selected database atomically for queries

## Verdict

The core design is sound and fixes a real bug at the right layer. The connection's selected database is server-side sticky state, and the old flow allowed another thread to interleave a `USE` between database selection and query execution. Moving the assertion inside the same `_lockConnection` acquisition as `mysql_real_query` is the correct fix, and the fallback behaviors are consistently conservative (unknown state → live `SELECT DATABASE()` check; any `USE`/`DROP DATABASE` → invalidate). I verified the scariest hole doesn't exist: the legacy `[SPMySQLConnection selectDatabase:]` routes through `queryString:` with a `USE` statement, so `recordSuccessfulQuery` invalidates the assertion state on every legacy database switch, and reconnects are caught by the connection-pointer + `mysql_thread_id` identity check. I'd approve after the performance issue below is addressed.

## Significant findings

**1. Performance regression in SQL import and custom-query batches (should fix before merge).**
`SASQLDatabaseContext.databaseName(afterSuccessfulQuery:...)` is called for *every* successful statement in `SPDataImport.importSQLFile:` and `SPCustomQuery.performQueriesTask:`, and it unconditionally runs `stripSQLComments`, which does `source.map { $0 }` (a `Character` array of the whole statement, ~16 bytes per character) and rebuilds the string with per-character `append`. The framework side guards this with a cheap first-non-whitespace-character check (`queryCouldChangeDatabaseContext`); the app side has **no guard at all**. A SQL dump full of multi-megabyte extended `INSERT` statements will pay a large transient allocation plus an O(n) character-by-character scan per statement — this can visibly slow imports, which is exactly the path this PR touches most. Please add the same cheap prefix guard before stripping (an `INSERT` starts with `I` and would skip entirely).

**2. The framework guard is coarser than it looks.**
`queryCouldChangeDatabaseContext`'s character set `"uUdD#-/"` matches every `UPDATE` (`U`) and `DELETE` (`D`), so all successful UPDATE/DELETE statements still pay the full comment-strip + regex in `recordSuccessfulQuery`. Checking the first *keyword* (case-insensitive `USE`/`DROP` prefix scan) instead of the first letter would eliminate nearly all remaining false positives on both sides. Row edits and chunked deletes in `SPTableContent` go through this on every statement.

**3. Misleading synthesized error for the assert-none case.**
`assertingDatabaseContext:nil` means "assert no database is selected," and when one *is* selected the code returns a synthesized 1046 "No database selected". That message states the opposite of what happened — a database is selected, just not the one the document expects. A user running a fully-qualified or `SHOW`-style query in a document with no selected database will get an error that reads as nonsense. Since MySQL has no way to *deselect* a database, failing is a defensible design choice, but the message should say something like "A database is unexpectedly selected on this connection."

**4. Invented MariaDB `[bracket]` identifier syntax.**
Both comment strippers and the `identifierPattern` regex treat `[...]` as a quoted-identifier form on MariaDB (`USE [new]]db]`, `DROP SCHEMA [App_]]DB]`), and tests cement it. That's SQL Server syntax — neither MySQL nor MariaDB supports bracket-quoted identifiers (MariaDB's `MSSQL` sql_mode doesn't enable it either). Practical risk is low (a bare `[` outside strings is nearly nonexistent in valid SQL), but the code models behavior the server doesn't have, and treating `[` as opening a quote could in principle mis-parse a valid query. It should be removed unless there's a MariaDB doc citation for it.

**5. Duplicated, already-diverging comment strippers.**
`stripSQLComments` + `shouldPreserveExecutableComment` now exist in two near-identical copies (framework `SADatabaseAssertion.swift`, app `SPCustomQuerySQLClassifier.swift`) with a real behavioral divergence: for a version gate whose digits fail `Int` parsing, the app side drops the body (`hasVersionGate && requiredVersion == nil → false`) while the framework side preserves it (its `if let` simply doesn't fire). The edge is absurd (a >19-digit gate), but it proves the copies are already drifting. Worth a shared implementation or at least a cross-referencing comment plus mirrored tests.

## Edge cases probed that hold up

- **Stale state via dynamic SQL** (`EXECUTE` of a prepared `DROP DATABASE`, which `recordSuccessfulQuery` can't see): the state wrongly stays `.selected`, the assertion skips reselection, and the query fails server-side with a genuine 1046. The failure mode is a visible error, never a silent wrong-database query. Acceptable.
- **Reconnect**: `databaseToRestore = [database copy]` at reconnect time now benefits from `setCurrentDatabaseFromQueryContext:` syncing `connection.database` (the property was already `readwrite, copy`, and atomic). The rewritten `refreshCurrentDatabase` only commits state inside the `![mySQLConnection queryErrored]` check, so transient failures don't wipe the selection.
- **Non-ASCII database names**: the live `@@character_set_client` lookup (rather than trusting session tracking) is right — servers can disable `session_track_system_variables`. The utf8mb4 select/restore sequence correctly prioritizes errors, and a restore failure deliberately overrides a successful selection so a leaked charset can't corrupt the caller's session.
- **API migration completeness**: grepping the base branch, every caller of `informationForTable:fromDatabase:nil`, `informationForView:`, and `getDatabaseDefaultCharacterSet`/`Collation` is updated in this diff; the old selectors are fully renamed so nothing can silently keep the stale behavior.
- **`SADatabaseScopedValueCache` thread safety**: not thread-safe itself, but all uses (including `reset` in `resetAllData`) sit under `@synchronized(charsetCollationLock)`. Fine. Asserting `nil` for the charset lookup when no database is selected correctly returns the server default.
- **`SPTableData` snapshot locals** (`selectedTableName` etc.) already exist on `main`; the PR just switches remaining `[tableListInstance tableName]` reads to them, which removes further races. Good.

## Smaller notes

- **Copy-view-to-another-database** behavior changed: `CREATE VIEW` now runs asserted against the *target* database, so unqualified table references in the view body resolve to the target schema (likely a creation error, or a view over the wrong tables). The old path was broken differently (`RENAME TABLE` can't move views across schemas), so this isn't a regression, but the new special-case branch in `_copyTable` deserves a manual test — the PR notes no manual import/export/copy UI pass was done against MariaDB.
- Each custom-query run and SQL import now issues one extra `SELECT @@lower_case_table_names` round-trip up front. Fine, but it could be cached per connection.
- `--` at end-of-input isn't treated as a comment (the stripper requires a character after the dashes); MySQL treats `-- <EOF>` as a comment. Classification-only impact, trivial.
- The `taskArguments` change in `performQueriesWithNoWarning:` incidentally fixes a latent bug where a nil `encodedCallbackMethod` truncated the old `dictionaryWithObjectsAndKeys:` list. Nice catch, worth a mention in the PR description.
- Project wiring is correct in both pbxproj files (new sources in app + both test targets; framework gets `MODULEMAP_FILE`, `SWIFT_VERSION`, `DEFINES_MODULE`). Note the framework now contains Swift with `BUILD_LIBRARY_FOR_DISTRIBUTION = YES` — module-stability interface generation is new build overhead, and it's worth confirming it's actually required here rather than cargo-culted; `@_implementationOnly import MySQLClient` is compatible with it, so it's safe either way. The `mainGroup` rename in the framework pbxproj is cosmetic churn.

## Test coverage

Strong overall: ~860 lines of framework tests including a live-MySQL integration class exercising shared-connection races and disabled session tracking, plus app-side tests for the classifier context tracking and the scoped cache. Gaps: no end-to-end coverage of the `SPCustomQuery` batch `USE`-following behavior (only the classifier is unit-tested), nothing exercises the copy-view-across-databases branch, and the bracket-identifier tests (finding 4) encode behavior that should probably be deleted.

## Bottom line

Well-engineered fix for a genuine concurrency bug, with unusually careful state invalidation and honest documentation of what's out of scope. Requesting changes for: the missing prefix guard in `SASQLDatabaseContext` (import performance), and preferably the misleading 1046 message and the bracket-identifier syntax before the tests calcify it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@Kaspik Thanks for the detailed pass. I addressed every actionable item from your changes-requested review and rebased the validation onto current `main` at head `02e222293`: exact low-cost `USE`/`DROP` guards, corrected no-database error text, removal of bracket identifiers, mirrored parser/gate/EOF coverage, no-whitespace quoted identifier handling, live custom-query batch and cross-database view-copy regressions, confirmed library-evolution requirement, removed project-file churn, documented the `taskArguments` fix, and cleared the new exporter warnings found in the final audit. Current results are 710/710 app tests passing with 60 expected skips, 27/27 framework tests passing, and a successful Release build. Would you please take a fresh review?

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by Jason-Morcos** [comment]
@Kaspik, the detailed review items are addressed on the latest signed head d26a85e5b, along with the subsequent Codex diagnostic-preservation finding. Successful USE/DROP statements now update assertion state precisely, and a live MySQL regression proves `DROP DATABASE IF EXISTS missing_schema; SHOW WARNINGS` retains warning 1008. Validation is 710 app tests passed (60 expected skips), 28/28 framework tests passed, and a zero-error Release build. Could you please re-review this latest head?

**Review by Kaspik — APPROVED** [review]

**Comment by Jason-Morcos** [comment]
@codex review especially making sure we are good with recent changes to main

**Comment by Jason-Morcos** [comment]
@codex review
